lms/djangoapps/badges/models.py

```diff
@@ -0,0 +1,328 @@
+        """
+        return self.backend.award(self, user, evidence_url=evidence_url)
+
+    def save(self, **kwargs):
+        """
+        Slugs must always be lowercase.
+        """
+        self.slug = self.slug and self.slug.lower()
+        self.issuing_component = self.issuing_component and self.issuing_component.lower()
+        super(BadgeClass, self).save(**kwargs)
+
+    class Meta(object):
+        app_label = "badges"
+        unique_together = (('slug', 'issuing_component', 'course_id'),)
+        verbose_name_plural = "Badge Classes"
+
+
+class BadgeAssertion(TimeStampedModel):
+    """
+    Tracks badges on our side of the badge baking transaction
+    """
+    user = models.ForeignKey(User)
+    badge_class = models.ForeignKey(BadgeClass)
+    data = JSONField()
+    backend = models.CharField(max_length=50)
+    image_url = models.URLField()
+    assertion_url = models.URLField()
+
+    def __unicode__(self):
+        return u"<{username} Badge Assertion for {slug} for {issuing_component}".format(
+            username=self.user.username, slug=self.badge_class.slug,
+            issuing_component=self.badge_class.issuing_component,
+        )
+
+    @classmethod
+    def assertions_for_user(cls, user, course_id=None):
+        """
+        Get all assertions for a user, optionally constrained to a course.
+        """
+        if course_id:
+            return cls.objects.filter(user=user, badge_class__course_id=course_id)
+        return cls.objects.filter(user=user)
+
+    class Meta(object):
+        app_label = "badges"
+
+
+# Abstract model doesn't index this, so we have to.
+BadgeAssertion._meta.get_field('created').db_index = True  # pylint: disable=protected-access
+
+
+class CourseCompleteImageConfiguration(models.Model):
+    """
+    Contains the icon configuration for badges for a specific course mode.
+    """
+    mode = models.CharField(
+        max_length=125,
+        help_text=_(u'The course mode for this badge image. For example, "verified" or "honor".'),
+        unique=True,
+    )
+    icon = models.ImageField(
+        # Actual max is 256KB, but need overhead for badge baking. This should be more than enough.
+        help_text=_(
+            u"Badge images must be square PNG files. The file size should be under 250KB."
+        ),
+        upload_to='course_complete_badges',
+        validators=[validate_badge_image]
+    )
+    default = models.BooleanField(
+        help_text=_(
+            u"Set this value to True if you want this image to be the default image for any course modes "
+            u"that do not have a specified badge image. You can have only one default image."
+        ),
+        default=False,
+    )
+
+    def __unicode__(self):
+        return u"<CourseCompleteImageConfiguration for '{mode}'{default}>".format(
+            mode=self.mode,
+            default=u" (default)" if self.default else u''
+        )
+
+    def clean(self):
```

Comment:
But now that I've looked up how others use the clean() method, I understand the intent and this is okay -- redacting my comment.